keystore.go

```diff
@@ -0,0 +1,161 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package libjvm
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/pavlo-v-chernykh/keystore-go/v4"
+	"golang.org/x/sys/unix"
+	"software.sslmate.com/src/go-pkcs12"
+)
+
+type Keystore interface {
+	Add(string, *pem.Block) error
+	Write() error
+}
+
+func DetectKeystore(location string) (Keystore, error) {
+	buf, err := os.ReadFile(location)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(buf) == 0 {
+		return nil, errors.New("empty keystore found")
+	}
+
+	if len(buf) > 3 && buf[0] == 0xFE && buf[1] == 0xED && buf[2] == 0xFE && buf[3] == 0xED {
+		return NewJKSKeystore(location, "changeit")
+	}
+
+	return NewPKCS12Keystore(location), nil
+}
+
+var _ Keystore = &JKSKeystore{}
+
+type JKSKeystore struct {
+	location string
+	password string
+	store    keystore.KeyStore
+}
+
+func NewJKSKeystore(location, password string) (*JKSKeystore, error) {
+	in, err := os.Open(location)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open %s\n%w", location, err)
+	}
+	defer in.Close()
+
+	ks := keystore.New(keystore.WithOrderedAliases())
+	if err := ks.Load(in, []byte(password)); err != nil {
+		return nil, fmt.Errorf("unable to decode keystore\n %w", err)
+	}
+	return &JKSKeystore{
+		location: location,
+		password: password,
+		store:    ks,
+	}, nil
+}
+
+func (k *JKSKeystore) Add(name string, b *pem.Block) error {
+	entry := keystore.TrustedCertificateEntry{
+		CreationTime: NormalizedDateTime,
+		Certificate: keystore.Certificate{
+			Type:    "X.509",
+			Content: b.Bytes,
+		},
+	}
+	if err := k.store.SetTrustedCertificateEntry(name, entry); err != nil {
+		return fmt.Errorf("unable to add trusted entry\n%w", err)
+	}
+	return nil
+}
+
+func (k *JKSKeystore) Write() error {
+	if unix.Access(k.location, unix.W_OK) != nil {
+		return nil
+	}
+
+	out, err := os.OpenFile(k.location, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open %s\n%w", k.location, err)
+	}
+	defer out.Close()
+
+	if err := k.store.Store(out, []byte(k.password)); err != nil {
+		return fmt.Errorf("unable to encode keystore\n%w", err)
+	}
+
+	return nil
+}
+
+func (k *JKSKeystore) Len() int {
+	return len(k.store.Aliases())
+}
+
+var _ Keystore = &PKCS12Keystore{}
+
+type PKCS12Keystore struct {
+	location string
+	entries  []pkcs12.TrustStoreEntry
+}
+
+func NewPKCS12Keystore(location string) *PKCS12Keystore {
+	return &PKCS12Keystore{
+		location: location,
+		entries:  []pkcs12.TrustStoreEntry{},
+	}
+}
```

Comment:
So this is for a password-less PKCS12 keystore, right?

If so...

1. Do you think we need to support passwords? I think all of the JVM vendors are using passwordless with their defaults, so from that perspective, it seems unnecessary. Unless you have a different JVM vendor, or someone's customized the default keystore 🤔 
2. If we don't support this now and if we were to need to support it at some point in the future, do you think we could add that to the existing code in a non-breaking way? If not, do you think we should rename this implementation PasswordLessPKCS12Keystore or something like that to leave room for a new implementation that does support passwords?

I don't think this is a huge issue, but just trying to be forward-thinking about this.

Reply:
I don't think that we need to support passwords, but I can check internally if there's some guidance. Renaming the implementation, to reflect that it does not use a password, seems reasonable to me. That way, we can introduce the functionality later without breaking anyone.